### PR TITLE
Refactor missused ViewModels, support automated UI tests and screenshots (EXPOSUREAPP-2946)

### PR DESCRIPTION
**Pull request by d4rken**
## Why is this PR so huge? :fearful:
Ripping the :adhesive_bandage:  off in one swoop saves us weeks of work.

The core idea is that we want to instantiate screens on their own to automate UI tests and take automated screenshots. This was impossible because there were too many dependencies that would have to be satisfied for each screen, sepperation was not possible because each screen accessed various other components via static access.

We had a lot of :ramen: code based on (`TracingViewModel`, `SubmissionViewModel` and `SettingsViewModel`) that was very interwoven in the apps structure. This PR removes the :ramen: and creates a one-way information flow. By that I mean that information flows from internal components (i.e. Bluetooth state) to external components (Fragments showing something) without extra roundtrips (i.e. we don't want individual fragments calling `refreshBluetooth`).

## What changed
* Reactive information flow for many elements, i.e. UI subscribes data, data changes, UI updates automatically, the UI no longer manually calls refresh in most cases.
* Fewer singleton access to allow easier testing.
* Most `***FormaterHelper` calls have been removed from the layout files, we had too much logic happening in XML files. Too difficult to add new features, too easy to break existing logic.
* We now have UI "state" objects. The viewmodels have a `Flow<State>` that can be observed, and if it changes the Fragment will update the layout via databinding. The remaining databinding code just sets the state variables to the relevant views. So now we have the best of both worlds. Not too complex logic in XML code. No extra boilerplate code in Fragments.
    * The states can easily be unit-tested on their own.
    * When writing instrumentation tests or doing  automated screenshots, we only need to create mock state objects to make the UI look a specific way.

The above patterns are something we should use throughout the app. This PR already refactors a lot but does not refactor everything. Only the places that were connected to the classes that needed to be refactored in relation to the `HomeFragment`... which is a lot :sweat_smile: 

## Testing
* Check the notification settings behavior in different states
* Check the tracing settings behavior in different states
* Check the behavior of the tracing status headline on the home screen in different states
* Check the risk card behavior in different states
* Check the submission card behavior in different states
* Check the risk details screen in different states
* Check submission flow

### CODE REVIEW

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/storage/TracingRepository.kt [hunk 1]
```diff
@@ -23,21 +24,26 @@ object TracingRepository {
...
      *
      * @see LocalData
      */
     fun refreshLastTimeDiagnosisKeysFetchedDate() {
-        lastTimeDiagnosisKeysFetched.postValue(LocalData.lastTimeDiagnosisKeysFromServerFetch())
+        internalLastTimeDiagnosisKeysFetched.value =
+            LocalData.lastTimeDiagnosisKeysFromServerFetch()
     }
 
+    // TODO shouldn't access this directly
```
Comment (ralfgehrer) [inline Corona-Warn-App/src/main/java/de/rki/coronawarnapp/storage/TracingRepository.kt:43]: Should go to top of class

**Review by chris-cwa — DISMISSED** [review]
This code is so clean you could perform open heart surgery on it!

File: Corona-Warn-App/src/deviceForTesters/java/de/rki/coronawarnapp/test/api/ui/TestForAPIFragment.kt [hunk 2]
```diff
@@ -258,34 +250,40 @@ class TestForAPIFragment : Fragment(R.layout.fragment_test_for_a_p_i),
...
                 }
             }
 
             buttonTracingDurationInRetentionPeriod.setOnClickListener {
-                tracingVM.viewModelScope.launch {
-                    showToast(TimeVariables.getActiveTracingDaysInRetentionPeriod().toString())
+                vm.launch {
+                    val daysInRetention =
+                        TimeVariables.getActiveTracingDaysInRetentionPeriod().toString()
+                    withContext(Dispatchers.Main) {
```
Comment (ralfgehrer) [inline Corona-Warn-App/src/deviceForTesters/java/de/rki/coronawarnapp/test/api/ui/TestForAPIFragment.kt:284]: Just a test fragment but maybe you want to inject the dispatcher provider anyway

File: Corona-Warn-App/src/deviceForTesters/java/de/rki/coronawarnapp/test/api/ui/TestForAPIFragment.kt [hunk 3]
```diff
@@ -258,34 +250,40 @@ class TestForAPIFragment : Fragment(R.layout.fragment_test_for_a_p_i),
...
-                tracingVM.viewModelScope.launch {
-                    showToast(
-                        TracingIntervalRepository.getDateRepository(requireContext()).getIntervals()
-                            .toString()
-                    )
+                vm.launch {
+                    val intervals = TracingIntervalRepository.getDateRepository(requireContext())
+                        .getIntervals()
+                        .toString()
+                    withContext(Dispatchers.Main) {
```
Comment (ralfgehrer) [inline Corona-Warn-App/src/deviceForTesters/java/de/rki/coronawarnapp/test/api/ui/TestForAPIFragment.kt:274]: Just a test fragment but maybe you want to inject the dispatcher provider anyway

File: Corona-Warn-App/src/deviceForTesters/java/de/rki/coronawarnapp/test/api/ui/TestForAPIFragment.kt [hunk 4]
```diff
@@ -258,34 +250,40 @@ class TestForAPIFragment : Fragment(R.layout.fragment_test_for_a_p_i),
...
-                tracingVM.viewModelScope.launch {
-                    showToast(
-                        ExposureSummaryRepository.getExposureSummaryRepository()
-                            .getExposureSummaryEntities().toString()
-                    )
+                vm.launch {
+                    val summary = ExposureSummaryRepository.getExposureSummaryRepository()
+                        .getExposureSummaryEntities().toString()
+
+                    withContext(Dispatchers.Main) {
```
Comment (ralfgehrer) [inline Corona-Warn-App/src/deviceForTesters/java/de/rki/coronawarnapp/test/api/ui/TestForAPIFragment.kt:257]: Just a test fragment but maybe you want to inject the dispatcher provider anyway

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/storage/SettingsRepository.kt [hunk 5]
```diff
@@ -23,65 +25,12 @@ class SettingsRepository @Inject constructor(
...
-     * Refresh notifications for test updates with the current shared preferences state.
-     *
-     * @see LocalData
-     */
-    fun refreshNotificationsTestEnabled() {
-        isNotificationsTestEnabled.value = LocalData.isNotificationsTestEnabled()
-    }
+    private val internalIsBackgroundPriorityEnabled = MutableStateFlow(false)
+    val isBackgroundPriorityEnabledFlow: Flow<Boolean> = internalIsBackgroundPriorityEnabled
+    val isBackgroundPriorityEnabled = internalIsBackgroundPriorityEnabled.asLiveData()
```
Comment (ralfgehrer) [inline Corona-Warn-App/src/main/java/de/rki/coronawarnapp/storage/SettingsRepository.kt:33]: I suppose the live data one will be deprecated eventually and you don't rename it to not change a million other occurrences? I'd propose to mark it as deprecated at least with a comment then.

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/storage/SettingsRepository.kt [hunk 6]
```diff
@@ -119,24 +50,33 @@ class SettingsRepository @Inject constructor(
         isBackgroundJobEnabled.postValue(value)
     }
 
+    private val internalIsManualKeyRetrievalEnabled = MutableStateFlow(true)
+    val isManualKeyRetrievalEnabledFlow: Flow<Boolean> = internalIsManualKeyRetrievalEnabled
+    val isManualKeyRetrievalEnabled = isManualKeyRetrievalEnabledFlow.asLiveData()
```
Comment (ralfgehrer) [inline Corona-Warn-App/src/main/java/de/rki/coronawarnapp/storage/SettingsRepository.kt:55]: same as above and for others in this repo

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/ui/tracing/settings/SettingsTracingFragment.kt [hunk 7]
```diff
@@ -141,7 +137,7 @@ class SettingsTracingFragment : Fragment(R.layout.fragment_settings_tracing),
             try {
                 if (InternalExposureNotificationClient.asyncIsEnabled()) {
                     InternalExposureNotificationClient.asyncStop()
-                    tracingViewModel.refreshIsTracingEnabled()
+//                    tracingViewModel.refreshIsTracingEnabled()
```
Comment (ralfgehrer) [inline Corona-Warn-App/src/main/java/de/rki/coronawarnapp/ui/tracing/settings/SettingsTracingFragment.kt:140]: If intended, why not delete it as with others.

**Review by ralfgehrer — DISMISSED** [review]
works well on my device (Pixel 1, Android 10)

**Comment by BMItter** [comment]
Tested on Pixel 2 (A11) & Nexus 5x (A8.1):
risk card & states, low risk tested. **i.o,** unknown, **i.o**
tracing state behavior **i.o.**, UI updates accordingly, **i.o**
tracing status headline on home screen, **i.o**
submission of test and UI check, states: pending **i.o**, negativ **i.o**, positiv **i.o.** , no test **i.o**
tested DefaultScanningSupport (new singleton), **i.o.**
Still doing some checks.

One short note: if risk detection is stopped by switch, after app restart initial card is showing "unknown risk", jumping to "riskdetection stopped" card. It might be the default behaviour in lack of peristancy, but I want to mention this.

File: Corona-Warn-App/src/deviceForTesters/java/de/rki/coronawarnapp/test/api/ui/TestForAPIFragment.kt (hunk 2 above)
Comment (d4rken) [reply Corona-Warn-App/src/deviceForTesters/java/de/rki/coronawarnapp/test/api/ui/TestForAPIFragment.kt:284]: The test fragment is not being tested currently, when we move this logic to a viewmodel, then we should use the injected dispatcher. While it's better, doesn't offer us much to do it now until it's under test.

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/storage/SettingsRepository.kt (hunk 5 above)
Comment (d4rken) [reply Corona-Warn-App/src/main/java/de/rki/coronawarnapp/storage/SettingsRepository.kt:33]: Ohh I like the idea, adding  `@Deprecated` :+1:

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/storage/SubmissionRepository.kt [hunk 8]
```diff
@@ -58,27 +60,75 @@ object SubmissionRepository {
...
+        TestResult.REDEEMED -> DeviceUIState.PAIRED_REDEEMED
+        TestResult.INVALID -> DeviceUIState.PAIRED_ERROR
+        null -> DeviceUIState.UNPAIRED
     }
 
     fun setTeletan(teletan: String) {
         LocalData.teletan(teletan)
     }
+
+    private val _uiStateError = MutableLiveData<Event<CwaWebException>>(null)
```
[the remark this answers is not in the document]
Comment (d4rken) [reply Corona-Warn-App/src/main/java/de/rki/coronawarnapp/storage/SubmissionRepository.kt:85]: Was copy pasted from previous code, but can fix.

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/ui/tracing/settings/SettingsTracingFragment.kt (hunk 7 above)
Comment (d4rken) [reply Corona-Warn-App/src/main/java/de/rki/coronawarnapp/ui/tracing/settings/SettingsTracingFragment.kt:140]: Was a reminder while refactoring, removed :+1:

**Review by ralfgehrer — APPROVED** [review]

**Review by BMItr — APPROVED** [review]

**Review by AlexanderAlferov — APPROVED** [review]

**Pull request merged**